filebeat/input/v2/plugin.go

```diff
@@ -0,0 +1,112 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package v2
+
+import (
+	"github.com/elastic/beats/heartbeat/scheduler"
+	"github.com/elastic/beats/libbeat/beat"
+	"github.com/elastic/beats/libbeat/common"
+	"github.com/elastic/beats/libbeat/logp"
+	"github.com/elastic/go-concert/chorus"
+)
+
+type Plugin struct {
+	Name        string
+	Constraints Constraint
+	Doc         string
+
+	// TODO: config schema info for config validation
+
+	// Configure configures an Input if possible. Returns an error if the
+	// configuration is false.
+	// The input return must not attempt to run or create any kind of connection yet.
+	// The input is only supposed to be a types place holder for the untyped
+	// configuration.
+	Configure func(log *logp.Logger, config *common.Config) (Input, error)
+}
+
+// Input defines the functionality all input plugins must provide.
+type Input interface {
+	InputTester
+	RunnerFactory
+}
+
+// InputTester provides functionality for testing aa feature its configuration without
+// having to run it for a long time.
+// The test is supposed to not take too much time. Packages running a tester
+// might run the test with a pre-configured timeout.
+type InputTester interface {
+	TestInput(closer *chorus.Closer, log *logp.Logger) error
```

Comment:
This could make sense at the plugin level. You also mention there the schema for config validation, where this would go beyond and actually test the config

Reply:
The Plugin level is just a definition of a feature we can provide. The `Input` type is no actual instance of an input collecting data, but a place-holder (interface type) for a configuration that has been read already read. It is used to create an instance later on.

The idea (as described already) is to build a complete typed configuration object in a first step, that then can be used by a command in the second step (no partially initialized 'beater'). For example the Filebeat struct would be:

```
type filebeat struct {
  Inputs []input.Input
  ...
}

func (fb *filebeat) Test(closer *chorus.Closer, log *logp.Logger) error {
  ...
  for _, input in fb.Inputs {
    err := input.TestInput(closer, log)
    ...
  }
  ...
}
```

With this in mind (long-term approach) each sub-command provided by beats may have multiple phases until the command can be executed. The read-config phase would be the common first-phase.